From 793b0a72dfd80556ad61d5ed822d7ec52fa0bc55 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Fri, 2 Jun 2023 16:50:47 +0800
Subject: [PATCH 402/780] [Huawei] blk-throttle: record hierarchy stats

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188826

--------------------------------

Support to account the numbers of io and io latency that is throttled.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 block/blk-throttle.c      | 4 ++++
 include/linux/blk_types.h | 5 ++++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/block/blk-throttle.c b/block/blk-throttle.c
index bcdd4e688060..996777cafc0e 100644
--- a/block/blk-throttle.c
+++ b/block/blk-throttle.c
@@ -14,6 +14,7 @@
 #include <linux/delay.h>
 #include "blk.h"
 #include "blk-cgroup-rwstat.h"
+#include "blk-io-hierarchy/stats.h"
 
 /* Max dispatch from a group in 1 round */
 #define THROTL_GRP_QUANTUM 8
@@ -1386,6 +1387,8 @@ static void blk_throtl_dispatch_work_fn(struct work_struct *work)
 			bio_list_add(&bio_list_on_stack, bio);
 	spin_unlock_irq(&q->queue_lock);
 
+	bio_list_hierarchy_end_io_acct(&bio_list_on_stack, STAGE_THROTTLE);
+
 	if (!bio_list_empty(&bio_list_on_stack)) {
 		blk_start_plug(&plug);
 		while ((bio = bio_list_pop(&bio_list_on_stack)))
@@ -2419,6 +2422,7 @@ bool blk_throtl_bio(struct bio *bio)
 	td->nr_queued[rw]++;
 	throtl_add_bio_tg(bio, qn, tg);
 	throttled = true;
+	bio_hierarchy_start_io_acct(bio, STAGE_THROTTLE);
 
 	/*
 	 * Update @tg's dispatch time and force schedule dispatch if @tg
diff --git a/include/linux/blk_types.h b/include/linux/blk_types.h
index 1b7867f22c68..d4030b3bba40 100644
--- a/include/linux/blk_types.h
+++ b/include/linux/blk_types.h
@@ -469,7 +469,10 @@ enum stat_group {
 };
 
 enum stage_group {
-	STAGE_RESERVE = 0,
+#ifdef CONFIG_BLK_DEV_THROTTLING
+	STAGE_THROTTLE,
+#endif
+	STAGE_RESERVE,
 	NR_STAGE_GROUPS,
 };
 
-- 
2.17.1

